src/typegate/src/runtimes/substantial/deno_context.ts

```diff
@@ -414,6 +416,53 @@ class RetryStrategy {
   }
 }
 
+
+class SubLogger {
+  constructor(private ctx: Context) {}
+
+  async #log(kind: "warn" | "error" | "info", ...args: unknown[]) {
+    await this.ctx.save(() => {
+      const prefix = `[${kind.toUpperCase()}: ${this.ctx.getRun().run_id}]`;
+      switch(kind) {
+        case "warn": {
+          console.warn(prefix, ...args);
+          break;
+        }
+        case "error": {
+          console.error(prefix,...args);
+          break;
+        }
+        default: {
+          console.info(prefix, ...args);
+          break;
+        }
+      }
+
+      const message = args.map((arg) => {
+        try {
+          return JSON.stringify(arg);
+        } catch(_) {
+          return String(arg);
```

Comment:
I had issues with circular objects. But I am pretty sure there are other instances when it will throw.

Reply:
Apparently it will also throw on [BigInt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify) and return undefined on other exotic cases (like functions).